### PR TITLE
Api8 - Map api

### PR DESCRIPTION
**Pull request by tyhdefu**
[SpongeAPI](https://github.com/SpongePowered/SpongeAPI/pull/2260) | SpongeCommon
I've ported the [API 7 Map api](https://github.com/SpongePowered/SpongeCommon/pull/2573) to api 8. It has a lot of changes in terms of sponge catalogs, but those amount to very little due to the cleaner system for API-8. As it stands I think the test cases I have written so far are working, but I have more tests I want to do before calling this fully ready to merge

### CODE REVIEW

File: src/main/java/org/spongepowered/common/data/provider/item/stack/MapInfoItemStackData.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import org.spongepowered.api.map.MapInfo;
+import org.spongepowered.common.SpongeCommon;
+import org.spongepowered.common.bridge.world.storage.MapDataBridge;
+import org.spongepowered.common.data.provider.DataProviderRegistrator;
+import org.spongepowered.common.util.Constants;
+
+import javax.annotation.Nullable;
+import java.util.Optional;
+
+public class MapInfoItemStackData {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/data/provider/item/stack/MapInfoItemStackData.java]: `final`

File: src/main/java/org/spongepowered/common/data/provider/map/MapInfoData.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+import org.spongepowered.common.bridge.ResourceKeyBridge;
+import org.spongepowered.common.bridge.world.storage.MapDataBridge;
+import org.spongepowered.common.data.provider.DataProviderRegistrator;
+import org.spongepowered.common.map.canvas.SpongeMapByteCanvas;
+import org.spongepowered.common.map.canvas.SpongeMapCanvas;
+import org.spongepowered.math.vector.Vector2i;
+
+import javax.annotation.Nullable;
+
+public class MapInfoData {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/data/provider/map/MapInfoData.java]: `final`

File: src/main/java/org/spongepowered/common/data/provider/item/stack/MapInfoItemStackData.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	// @formatter:off
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/data/provider/item/stack/MapInfoItemStackData.java:45]: private constructor

File: src/main/java/org/spongepowered/common/inventory/custom/SpongeInventoryMenu.java [hunk 4]
```diff
@@ -259,12 +259,12 @@ public boolean onClick(int slotId, int dragType, ClickType clickTypeIn, PlayerEn
                         // * Throwing one item on the cursor by clicking outside the inventory window.
                         // */
                         //public static final org.spongepowered.api.item.inventory.menu.ClickType
-                        //        CLICK_THROW_ONE = DummyObjectProvider.createFor(org.spongepowered.api.item.inventory.menu.ClickType.class, "click_throw_one");
+                        //        CLICK_THROW_ONE = Sponge.getRegistry().getCatalogRegistry().provideSupplier(org.spongepowered.api.item.inventory.menu.ClickType.class, "click_throw_one");
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/inventory/custom/SpongeInventoryMenu.java]: This shouldn't be in this PR.

File: src/main/java/org/spongepowered/common/data/provider/map/MapInfoDataProviders.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.data.provider.map;
+
+import org.spongepowered.common.data.provider.DataProviderRegistratorBuilder;
+import org.spongepowered.common.util.Constants;
+
+public class MapInfoDataProviders extends DataProviderRegistratorBuilder {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/data/provider/map/MapInfoDataProviders.java]: `final`

File: src/main/java/org/spongepowered/common/map/MapUtil.java [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+     * Returns MapInfo of newly created map, if the event was not cancelled.
+     * @param cause Cause of the event
+     * @return MapInfo if event was not cancelled
+     */
+    public static Optional<MapInfo> fireCreateMapEvent(final Cause cause) {
+        return MapUtil.fireCreateMapEvent(cause, Collections.emptySet());
+    }
+
+    public static Optional<MapInfo> fireCreateMapEvent(final Cause cause, Set<Value<?>> values) {
+        LogManager.getLogger().info("in fireCreateMapEvent!");
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/map/MapUtil.java]: Debugging line missed?

File: src/main/java/org/spongepowered/common/map/MapUtil.java [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+    /**
+     * Returns MapInfo of newly created map, if the event was not cancelled.
+     * @param cause Cause of the event
+     * @return MapInfo if event was not cancelled
+     */
+    public static Optional<MapInfo> fireCreateMapEvent(final Cause cause) {
+        return MapUtil.fireCreateMapEvent(cause, Collections.emptySet());
+    }
+
+    public static Optional<MapInfo> fireCreateMapEvent(final Cause cause, Set<Value<?>> values) {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/map/MapUtil.java]: This should be in `SpongeCommonEventFactory`

File: src/main/java/org/spongepowered/common/map/MapUtil.java [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+import org.spongepowered.common.util.Constants;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+public final class MapUtil {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/map/MapUtil.java]: This should be in `org.spongepowered.common.util`

File: src/main/java/org/spongepowered/common/map/SpongeMapStorage.java [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+import java.util.Set;
+import java.util.UUID;
+
+/**
+ * Although minecraft no longer has a "Map Storage", it makes sense
+ * as a concept to provide the plugin creators with a single place
+ * to handle maps, instead of putting the methods at overworld,
+ * or at server level.
+ */
+public class SpongeMapStorage implements MapStorage {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/map/SpongeMapStorage.java]: `final`

File: src/main/java/org/spongepowered/common/map/canvas/SpongeEmptyCanvas.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+import java.awt.Color;
+import java.awt.Graphics;
+import java.awt.Image;
+import java.awt.image.BufferedImage;
+import java.util.Arrays;
+import java.util.Collections;
+
+// This class basically means that we don't have to create tons of huge byte arrays
+// While doing stuff like create SpongeValues for keys.
+public class SpongeEmptyCanvas implements SpongeMapCanvas {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/map/canvas/SpongeEmptyCanvas.java]: `final`

File: src/main/java/org/spongepowered/common/map/canvas/SpongeMapByteCanvas.java [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+import org.spongepowered.common.util.Constants;
+
+import java.awt.Color;
+import java.awt.Image;
+import java.awt.image.BufferedImage;
+
+import static com.google.common.base.Preconditions.checkState;
+
+// MapCanvas backed by a byte array
+public class SpongeMapByteCanvas implements SpongeMapCanvas {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/map/canvas/SpongeMapByteCanvas.java]: `final`

File: src/main/java/org/spongepowered/common/map/canvas/SpongeEmptyCanvas.java (hunk 10 above)
Comment (Zidane) [reply src/main/java/org/spongepowered/common/map/canvas/SpongeEmptyCanvas.java]: Additionally, `final` parameters.

File: src/main/java/org/spongepowered/common/map/canvas/SpongeMapByteCanvas.java (hunk 11 above)
Comment (Zidane) [reply src/main/java/org/spongepowered/common/map/canvas/SpongeMapByteCanvas.java]: Additionally, `final` parameters.

File: src/main/java/org/spongepowered/common/map/canvas/SpongeMapByteCanvas.java [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    @Override
+    public DataContainer toContainer() {
+        return DataContainer.createNew().set(Constants.Map.MAP_CANVAS, Bytes.asList(canvas));
+    }
+
+    @Override
+    public MapColor getColor(int x, int y) {
+        checkState(MapUtil.isInCanvasBounds(x), "x is out of bounds");
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/map/canvas/SpongeMapByteCanvas.java]: We're trying to move away from Guava as much as possible.

Do the if check and then throw the illegal state yourself.

File: src/main/java/org/spongepowered/common/map/canvas/SpongeMapCanvasBuilder.java [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+import java.awt.Graphics2D;
+import java.awt.Image;
+import java.awt.image.BufferedImage;
+import java.awt.image.DataBufferInt;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+public class SpongeMapCanvasBuilder implements MapCanvas.Builder {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/map/canvas/SpongeMapCanvasBuilder.java]: `final`

Additionally, `final` parameters.

File: src/main/java/org/spongepowered/common/map/color/SpongeMapColorType.java [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.map.color;
+
+import net.minecraft.block.material.MaterialColor;
+import org.spongepowered.api.ResourceKey;
+import org.spongepowered.api.map.color.MapColorType;
+import org.spongepowered.api.util.Color;
+
+public class SpongeMapColorType implements MapColorType {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/map/color/SpongeMapColorType.java]: This is a catalog, extend `SpongeCatalogType`

File: src/main/java/org/spongepowered/common/map/color/SpongeMapShade.java [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.map.color;
+
+import org.spongepowered.api.ResourceKey;
+import org.spongepowered.api.map.color.MapShade;
+import org.spongepowered.common.util.Constants;
+
+public class SpongeMapShade implements MapShade {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/map/color/SpongeMapShade.java]: This is a CatalogType, extend SpongeCatalogType.

File: src/main/java/org/spongepowered/common/map/decoration/SpongeMapDecorationType.java [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.map.decoration;
+
+import net.minecraft.world.storage.MapDecoration;
+import org.spongepowered.api.ResourceKey;
+import org.spongepowered.api.map.decoration.MapDecorationType;
+
+public class SpongeMapDecorationType implements MapDecorationType {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/map/decoration/SpongeMapDecorationType.java:30]: extend SpongeCatalogType.

File: src/main/java/org/spongepowered/common/map/decoration/orientation/SpongeMapDecorationOrientation.java [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.map.decoration.orientation;
+
+import org.spongepowered.api.ResourceKey;
+import org.spongepowered.api.map.decoration.orientation.MapDecorationOrientation;
+
+public class SpongeMapDecorationOrientation implements MapDecorationOrientation {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/map/decoration/orientation/SpongeMapDecorationOrientation.java:29]: extend SpongeCatalogType

File: src/main/java/org/spongepowered/common/registry/builtin/sponge/MapColorTypeStreamGenerator.java [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+import java.util.Optional;
+import java.util.stream.Stream;
+
+public class MapColorTypeStreamGenerator {
+	private static final Map<Integer, MapColorType> COLOR_INDEX_MAP = new HashMap<>();
+	private static final Logger logger = LogManager.getLogger(MapColorTypeStreamGenerator.class.getSimpleName());
+
+	public static Stream<MapColorType> stream() {
+		///////////////////////// ORDER MATTERS HERE ////////////////////
+		return Stream.of(
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/registry/builtin/sponge/MapColorTypeStreamGenerator.java]: Correct the formatting here please to match the others.

File: src/main/java/org/spongepowered/common/registry/builtin/sponge/MapColorTypeStreamGenerator.java [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+				addToRegistry("purple_terracotta", 		MaterialColor.PURPLE_TERRACOTTA),
+				addToRegistry("blue_terracotta", 		MaterialColor.BLUE_TERRACOTTA),
+				addToRegistry("brown_terracotta", 		MaterialColor.BROWN_TERRACOTTA),
+				addToRegistry("green_terracotta", 		MaterialColor.GREEN_TERRACOTTA),
+				addToRegistry("red_terracotta", 		MaterialColor.RED_TERRACOTTA),
+				addToRegistry("black_terracotta", 		MaterialColor.BLACK_TERRACOTTA)
+		);
+	}
+
+	private static MapColorType addToRegistry(final String id, final MaterialColor color) {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/registry/builtin/sponge/MapColorTypeStreamGenerator.java]: `newType`

File: src/main/java/org/spongepowered/common/registry/builtin/sponge/MapColorTypeStreamGenerator.java [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+
+public class MapColorTypeStreamGenerator {
+	private static final Map<Integer, MapColorType> COLOR_INDEX_MAP = new HashMap<>();
+	private static final Logger logger = LogManager.getLogger(MapColorTypeStreamGenerator.class.getSimpleName());
+
+	public static Stream<MapColorType> stream() {
+		///////////////////////// ORDER MATTERS HERE ////////////////////
+		return Stream.of(
+				addToRegistry("air", 					MaterialColor.AIR),
+				addToRegistry("grass", 					MaterialColor.GRASS),
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/registry/builtin/sponge/MapColorTypeStreamGenerator.java]: Fully qualify static access.

File: src/mixins/java/org/spongepowered/common/mixin/api/mcp/map/MapDecorationMixin_API.java [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+import org.spongepowered.common.bridge.world.storage.MapDecorationBridge;
+import org.spongepowered.common.map.MapUtil;
+import org.spongepowered.common.map.decoration.orientation.SpongeMapDecorationOrientation;
+import org.spongepowered.common.registry.builtin.sponge.MapDecorationOrientationStreamGenerator;
+import org.spongepowered.common.registry.builtin.sponge.MapDecorationTypeStreamGenerator;
+import org.spongepowered.common.util.Constants;
+import org.spongepowered.math.vector.Vector2i;
+
+@Mixin(MapDecoration.class)
+public class MapDecorationMixin_API implements org.spongepowered.api.map.decoration.MapDecoration {
```
Comment (Zidane) [inline src/mixins/java/org/spongepowered/common/mixin/api/mcp/map/MapDecorationMixin_API.java]: Mixin classes are `abstract`

File: src/mixins/java/org/spongepowered/common/mixin/core/item/ItemEmptyMapMixin.java [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+    @Redirect(method = "onItemRightClick",
+            at = @At(value = "INVOKE",
+                    target = "Lnet/minecraft/item/FilledMapItem;setupNewMap(Lnet/minecraft/world/World;IIBZZ)Lnet/minecraft/item/ItemStack;",
+                    ordinal = 0))
+    private ItemStack impl$createMapWithSpongeData(World worldIn, int worldX, int worldZ, byte scale, boolean trackingPosition, boolean unlimitedTracking,
+                                                   World worldIn2, PlayerEntity playerIn, Hand handIn) {
+        final Player player = (Player) playerIn;
+
+        try (final CauseStackManager.StackFrame frame = PhaseTracker.getCauseStackManager().pushCauseFrame()) {
+            frame.pushCause(player);
```
Comment (Zidane) [inline src/mixins/java/org/spongepowered/common/mixin/core/item/ItemEmptyMapMixin.java]: Player ticks/Packet processes already push the player on the cause stack, not needed.

File: src/main/java/org/spongepowered/common/registry/builtin/sponge/MapColorTypeStreamGenerator.java (hunk 18 above)
Comment (tyhdefu) [reply src/main/java/org/spongepowered/common/registry/builtin/sponge/MapColorTypeStreamGenerator.java]: Not sure exactly what you want me to do here.

**Comment by tyhdefu** [comment]
Ready for merge as far as I know. It looks like i've failed terribly at rebasing on multiple occassions - Maybe the commits should be squashed or something?

## Example added data:

### Vanilla map files
```
"": {
    data: {
        zCenter: -128
        unlimitedTracking: 1B
        trackingPosition: 1B
        frames: []
        scale: 0B
        locked: 0B
        dimension: "minecraft:overworld"
        banners: []
        xCenter: 0
        colors: bytes(..... lots of bytes ....)
    }
    SpongeData: {
        Decorations: [
            {
                rot: 6
                x: 112
                z: -128
                id: sponge-12aa55ac-21b1-48c7-b118-729f05540160
                type: 2
            },
            {
                rot: 4
                x: -64
                z: -128
                id: sponge-b5c88721-1682-452b-bd6c-cfb787c89626
                type: 2
            },
            {
                rot: 14
                x: 64
                z: -128
                id: sponge-f33ac1a1-ec14-4c47-b17b-4cf29ded5290
                type: 2
            },
            {
                rot: 15
                x: -112
                z: -128
                id: sponge-c9d4a8ac-2548-4520-9afc-9b0ef44ea84c
                type: 2
            },
            {
                rot: 5
                x: -48
                z: -128
                id: sponge-de420aed-35d8-4353-bc24-c9eedea72f6e
                type: 2
            },
            {
                rot: 1
                x: 32
                z: -128
                id: sponge-d3e36074-41b9-433a-9b9d-6fc6738fb396
                type: 2
            },
            {
                rot: 8
                x: 16
                z: -128
                id: sponge-f20fdb6d-67fa-4657-9ae4-1dce43eaf005
                type: 2
            },
            {
                rot: 0
                x: 80
                z: -128
                id: sponge-a03ec7db-d9e1-46a2-8f81-8fd65ff82293
                type: 2
            },
            {
                rot: 7
                x: -32
                z: -128
                id: sponge-fd329ae6-3afe-4e46-9a86-d125ff97e6ab
                type: 2
            },
            {
                rot: 12
                x: -80
                z: -128
                id: sponge-850890ec-9ad7-4c50-96ce-3767aff2413e
                type: 2
            },
            {
                rot: 13
                x: -128
                z: -128
                id: sponge-8eb6a390-8d49-4c0a-a93d-cd549bce6824
                type: 2
            },
            {
                rot: 10
                x: 48
                z: -128
                id: sponge-87d95261-9e6a-43dd-baa2-70c604a01566
                type: 2
            },
            {
                rot: 11
                x: -96
                z: -128
                id: sponge-cf3560a4-fabf-4421-affa-2ce5bd1b6346
                type: 2
            },
            {
                rot: 9
                x: -16
                z: -128
                id: sponge-6af6b8af-3bb3-47a3-8705-3b73a82cde67
                type: 2
            },
            {
                rot: 2
                x: 0
                z: -128
                id: sponge-28b8bc9d-a794-4b97-b0c8-86e0550aeced
                type: 2
            },
            {
                rot: 3
                x: 96
                z: -128
                id: sponge-d19dcc60-f4c9-4b01-8a84-4c18ca315cc5
                type: 2
            },
        ]
    }
    DataVersion: 2586
}
```

### level.dat
```
"": {
    SpongeData: {
        MapUUIDs: {
            "0": ints(
                -757778355,
                1428900935,
                -2047149444,
                1212142356
            )
            "1": ints(
                -409981466,
                -31439851,
                -1397946763,
                -1651835627
            )
            "2": ints(
                -1091813237,
                169427106,
                -1520482429,
                -866096061
            )
            "3": ints(
                1497803196,
                1515472867,
                -1450773522,
                568453198
            )
            "4": ints(
                1395764055,
                -1955118798,
                -1161208903,
                -940021831
            )
        }
        UUID: ints(
            1961410109,
            2133937682,
            -1820582295,
            -843544624
        )
    }
    Data: {
        WanderingTraderSpawnChance: 75
        BorderCenterZ: 0.0
        Difficulty: 1B
        BorderSizeLerpTime: 0L
        raining: 0B
        Time: 283070L
        GameType: 0
        ServerBrands: [sponge,]
        BorderCenterX: 0.0
        BorderDamagePerBlock: 0.2
        BorderWarningBlocks: 5.0
        WorldGenSettings: {
            bonus_chest: 0B
            seed: -5706222210593531649L
            generate_features: 1B
            dimensions: {
                "minecraft:overworld": {
                    #sponge: {
                        performs_spawn_logic: 0B
                        load_on_startup: 1B
                    }
                    generator: {
                        settings: "minecraft:overworld"
                        seed: -5706222210593531649L
                        biome_source: {
                            large_biomes: 0B
                            seed: -5706222210593531649L
                            #sponge: {
                                biomes: [
                                    "minecraft:ocean",
                                    "minecraft:plains",
                                    "minecraft:desert",
                                    "minecraft:mountains",
                                    "minecraft:forest",
                                    "minecraft:taiga",
                                    "minecraft:swamp",
                                    "minecraft:river",
                                    "minecraft:frozen_ocean",
                                    "minecraft:frozen_river",
                                    "minecraft:snowy_tundra",
                                    "minecraft:snowy_mountains",
                                    "minecraft:mushroom_fields",
                                    "minecraft:mushroom_field_shore",
                                    "minecraft:beach",
                                    "minecraft:desert_hills",
                                    "minecraft:wooded_hills",
                                    "minecraft:taiga_hills",
                                    "minecraft:mountain_edge",
                                    "minecraft:jungle",
                                    "minecraft:jungle_hills",
                                    "minecraft:jungle_edge",
                                    "minecraft:deep_ocean",
                                    "minecraft:stone_shore",
                                    "minecraft:snowy_beach",
                                    "minecraft:birch_forest",
                                    "minecraft:birch_forest_hills",
                                    "minecraft:dark_forest",
                                    "minecraft:snowy_taiga",
                                    "minecraft:snowy_taiga_hills",
                                    "minecraft:giant_tree_taiga",
                                    "minecraft:giant_tree_taiga_hills",
                                    "minecraft:wooded_mountains",
                                    "minecraft:savanna",
                                    "minecraft:savanna_plateau",
                                    "minecraft:badlands",
                                    "minecraft:wooded_badlands_plateau",
                                    "minecraft:badlands_plateau",
                                    "minecraft:warm_ocean",
                                    "minecraft:lukewarm_ocean",
                                    "minecraft:cold_ocean",
                                    "minecraft:deep_warm_ocean",
                                    "minecraft:deep_lukewarm_ocean",
                                    "minecraft:deep_cold_ocean",
                                    "minecraft:deep_frozen_ocean",
                                    "minecraft:sunflower_plains",
                                    "minecraft:desert_lakes",
                                    "minecraft:gravelly_mountains",
                                    "minecraft:flower_forest",
                                    "minecraft:taiga_mountains",
                                    "minecraft:swamp_hills",
                                    "minecraft:ice_spikes",
                                    "minecraft:modified_jungle",
                                    "minecraft:modified_jungle_edge",
                                    "minecraft:tall_birch_forest",
                                    "minecraft:tall_birch_hills",
                                    "minecraft:dark_forest_hills",
                                    "minecraft:snowy_taiga_mountains",
                                    "minecraft:giant_spruce_taiga",
                                    "minecraft:giant_spruce_taiga_hills",
                                    "minecraft:modified_gravelly_mountains",
                                    "minecraft:shattered_savanna",
                                    "minecraft:shattered_savanna_plateau",
                                    "minecraft:eroded_badlands",
                                    "minecraft:modified_wooded_badlands_plateau",
                                    "minecraft:modified_badlands_plateau",
                                ]
                            }
                            type: "minecraft:vanilla_layered"
                        }
                        type: "minecraft:noise"
                    }
                    type: "minecraft:overworld"
                }
                "minecraft:the_nether": {
                    #sponge: {
                        performs_spawn_logic: 0B
                        load_on_startup: 1B
                    }
                    generator: {
                        settings: "minecraft:nether"
                        seed: -5706222210593531649L
                        biome_source: {
                            seed: -5706222210593531649L
                            preset: "minecraft:nether"
                            type: "minecraft:multi_noise"
                        }
                        type: "minecraft:noise"
                    }
                    type: "minecraft:the_nether"
                }
                "minecraft:the_end": {
                    #sponge: {
                        performs_spawn_logic: 0B
                        load_on_startup: 1B
                    }
                    generator: {
                        settings: "minecraft:end"
                        seed: -5706222210593531649L
                        biome_source: {
                            seed: -5706222210593531649L
                            type: "minecraft:the_end"
                        }
                        type: "minecraft:noise"
                    }
                    type: "minecraft:the_end"
                }
            }
        }
        DragonFight: {}
        BorderSizeLerpTarget: 6.0E7
        Version: {
            Snapshot: 0B
            Id: 2586
            Name: "1.16.5"
        }
        DayTime: 283070L
        initialized: 1B
        WasModded: 1B
        allowCommands: 0B
        WanderingTraderSpawnDelay: 20400
        CustomBossEvents: {}
        GameRules: {
            doFireTick: "true"
            maxCommandChainLength: "65536"
            fireDamage: "true"
            reducedDebugInfo: "false"
            disableElytraMovementCheck: "false"
            announceAdvancements: "true"
            drowningDamage: "true"
            commandBlockOutput: "true"
            forgiveDeadPlayers: "true"
            doMobSpawning: "true"
            maxEntityCramming: "24"
            disableRaids: "false"
            doWeatherCycle: "true"
            doDaylightCycle: "true"
            showDeathMessages: "true"
            doTileDrops: "true"
            universalAnger: "false"
            doInsomnia: "true"
            doImmediateRespawn: "false"
            naturalRegeneration: "true"
            doMobLoot: "true"
            fallDamage: "true"
            keepInventory: "false"
            doEntityDrops: "true"
            doLimitedCrafting: "false"
            mobGriefing: "true"
            randomTickSpeed: "3"
            spawnRadius: "10"
            doTraderSpawning: "true"
            logAdminCommands: "true"
            spectatorsGenerateChunks: "true"
            sendCommandFeedback: "true"
            doPatrolSpawning: "true"
        }
        SpawnY: 64
        rainTime: 62917
        thunderTime: 2608
        SpawnZ: -192
        hardcore: 0B
        WanderingTraderId: ints(
            -359330881,
            512377501,
            -1599446386,
            -422000034
        )
        DifficultyLocked: 0B
        SpawnX: 16
        clearWeatherTime: 0
        thundering: 1B
        SpawnAngle: 0.0F
        version: 19133
        BorderSafeZone: 5.0
        LastPlayed: 1614780891287L
        BorderWarningTime: 15.0
        ScheduledEvents: []
        LevelName: world
        BorderSize: 6.0E7
        DataVersion: 2586
        DataPacks: {
            Enabled: [vanilla,]
            Disabled: []
        }
    }
}
```

**Comment by Faithcaio** [comment]
merged in https://github.com/SpongePowered/Sponge/commit/a13921534c73e0628b50b35405841454811b861c

**Pull request closed without merging**